lib/Synchronise/Network/Server.hs

```diff
@@ -212,9 +230,137 @@ notify (RequestChange note) = do
             throwError UnknownKeyError
         Just _ ->
             liftIO . addWork store . WorkNotify $ ForeignKey ent_name src_name fid
+
     return ResponseChange
 
+
 --------------------------------------------------------------------------------
+
+-- * Asynchronous Server Workers
+
+-- | Get a work item from the work queue, apply a constant backoff if there is
+--   nothing in the queue.
+--
+--   This function must not
+getWorkBackoff :: Store store => store -> IO (WorkItemID, WorkItem)
+getWorkBackoff store = do
+  work <- getWork store
+  case work of
+    Nothing -> threadDelay 50000 >> getWorkBackoff store
+    Just x  -> return x
+
+-- | A worker for the synchronised server.
+--   These workers cannot die, they simply log any errors and keep going.
+--
+worker :: Store store => store -> Configuration -> IO ()
+worker store cfg = go
+  where go = do
+          (_, item) <- liftIO $ getWorkBackoff store
+          case item of
+            WorkNotify fk -> do
+              liftIO . debugM logName $ "Processing a notifcation: " <> show fk
+              processNotification store cfg fk
+            WorkApplyPatch did diff -> do
+              liftIO . debugM logName $ "Processing a diff: " <> show did
+              processDiff did diff
+          go
+
+-- notifications
+
+processNotification :: Store store => store -> Configuration -> ForeignKey -> IO ()
+processNotification store cfg fk@(ForeignKey{..}) = do
+  let ds = do es <- M.lookup fkEntity (configEntities cfg)
+              _  <- M.lookup fkSource (entitySources es)
+              M.lookup fkSource (entitySources es)
+  case ds of
+    Nothing         -> liftIO . errorM logName $ "Unknown key in workqueue: " <> show fk
```

Comment:
We might want this to be higher than `ERROR`; it means we accepted work we can't do. Either there's a super bad bug or we've changed the config and broken it.